### PR TITLE
Feature/redux

### PR DESCRIPTION
**Pull request by limanovic**

### CODE REVIEW

File: src/app/components/Input/AddTask.tsx [hunk 1]
```diff
@@ -1,104 +1,95 @@
...
-                alert('Task cannot be blank');
-            } else {
-                setTasks(updatedTasks);
-                setNewTask('');
-                setEditTask(null);
-            }
-            setOpenModal(false);
-        }
+    const openSaveModal = () => {
+        if (changedName.length < 1 || changedName.trim() == '') alert('Task cannot be empty');
```
[the remark this answers is not in the document]
Comment (asapMaki) [reply src/app/components/Input/AddTask.tsx:36]: !changedName.trim()

[the remark this answers is not in the document]
Comment (asapMaki) [reply src/app/components/Input/AddTask.tsx:36]: !changedName.length

File: src/app/components/Redux/reducers.tsx [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { combineReducers } from '@reduxjs/toolkit';
+import taskReducer from './addTaskSlice';
```
Comment (asapMaki) [inline src/app/components/Redux/reducers.tsx:2]: create folder tasks and file in it called slice.ts

File: src/app/components/Task/Task.tsx [hunk 3]
```diff
@@ -1,53 +1,53 @@
 import { Button } from '@mui/material';
 import React, { useState } from 'react';
 import ConfirmModal from '../ConfirmModal';
+import { useDispatch } from 'react-redux';
+import { removeTask, editTask } from '../Redux/addTaskSlice';
 
 type Task = {
     id: number;
     name: string;
+    isEditing: string | null;
```
Comment (asapMaki) [inline src/app/components/Task/Task.tsx:10]: can be isEditing?:

File: src/app/components/Task/Task.tsx [hunk 4]
```diff
@@ -1,53 +1,53 @@
...
-    tasks,
-    setTasks,
-}: {
-    task: Task;
-    editTaskName: (task: Task) => void;
-    tasks: Task[];
-    setTasks: React.Dispatch<React.SetStateAction<Task[]>>;
-}) {
+export default function Task({ task }: { task: Task }) {
     const [removeModal, setRemoveModal] = useState<boolean>(false);
```
Comment (asapMaki) [inline src/app/components/Task/Task.tsx:13]: reanme to modalOpened

File: src/app/components/Input/AddTask.tsx [hunk 5]
```diff
@@ -1,104 +1,95 @@
...
+
+const AddTask = () => {
+    const dispatch = useDispatch();
+
+    const inputRef = useRef<HTMLInputElement>(null);
+    let tasks = useSelector((state: RootState) => state.tasks);
+    const [taskName, setTaskName] = useState({ name: '', id: 0, isEditing: null });
+    const task = tasks.find((task: Task) => task.name === task.isEditing);
+    const [changedName, setChangedName] = useState<string>('');
+    const [saveModal, setSaveModal] = useState<boolean>(false);
```
Comment (asapMaki) [inline src/app/components/Input/AddTask.tsx:25]: rename to modalOpened, unless you have multiple modals

**Review by asapMaki — CHANGES_REQUESTED** [review]

File: src/app/components/AddTask.tsx [hunk 6]
```diff
@@ -1,48 +1,40 @@
...
 
     const inputRef = useRef<HTMLInputElement>(null);
-    let tasks = useSelector((state: RootState) => state.tasks);
-    const [taskName, setTaskName] = useState({ name: '', id: 0, isEditing: null });
-    const task = tasks.find((task: Task) => task.name === task.isEditing);
+    const [taskName, setTaskName] = useState({ name: '', id: 0 });
     const [changedName, setChangedName] = useState<string>('');
-    const [saveModal, setSaveModal] = useState<boolean>(false);
+    const [modalOpened, setmodalOpened] = useState<boolean>(false);
+    let tasks = useSelector((state: RootState) => state.tasks);
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:15]: https://redux-toolkit.js.org/tutorials/typescript
Check in mobile repo for usage

File: src/app/components/AddTask.tsx [hunk 7]
```diff
@@ -1,48 +1,40 @@
...
-    id: number;
-    name: string;
-    isEditing: string | null;
-};
-
-interface RootState {
-    tasks: Task[];
-}
+import { addTask, saveTask } from './Redux/tasks/slice';
+import { TaskType, RootState } from './Redux/types';
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:6]: type Task = ...

import { Task as TaskType ...

Comment (asapMaki) [reply src/app/components/AddTask.tsx:6]: Talk with Davud

File: src/app/components/AddTask.tsx [hunk 8]
```diff
@@ -1,48 +1,40 @@
...
 
 const AddTask = () => {
     const dispatch = useDispatch();
 
     const inputRef = useRef<HTMLInputElement>(null);
-    let tasks = useSelector((state: RootState) => state.tasks);
-    const [taskName, setTaskName] = useState({ name: '', id: 0, isEditing: null });
-    const task = tasks.find((task: Task) => task.name === task.isEditing);
+    const [taskName, setTaskName] = useState({ name: '', id: 0 });
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:12]: in constants create initial task object to use throughout the apps

File: src/app/components/AddTask.tsx [hunk 9]
```diff
@@ -1,48 +1,40 @@
...
+    let tasks = useSelector((state: RootState) => state.tasks);
+    const task = tasks.find((task: TaskType) => task.name === task.isEditing);
 
     const handleAddTask = (e: React.FormEvent<HTMLFormElement>) => {
         e.preventDefault();
-        dispatch(addTask(taskName));
-        setTaskName({ name: '', id: 0, isEditing: null });
+        if (!taskName || !taskName.name.trim()) alert('Task cannot be empty');
+        else dispatch(addTask(taskName));
+        setTaskName({ name: '', id: 0 });
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:22]: initial task constant

File: src/app/components/AddTask.tsx [hunk 10]
```diff
@@ -1,48 +1,40 @@
...
     };
     const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
         setChangedName(e.target.value);
     };
-    const openSaveModal = () => {
-        if (changedName.length < 1 || changedName.trim() == '') alert('Task cannot be empty');
-        else setSaveModal(true);
+    const openmodalOpened = () => {
+        if (!changedName || !changedName.trim()) alert('Task cannot be empty');
+        else setmodalOpened(true);
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:29]: setModalOpened*

File: src/app/components/AddTask.tsx [hunk 11]
```diff
@@ -1,48 +1,40 @@
...
     };
     const inputChangeNotConfirmed = () => {
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:31]: maybe name cancel ?

File: src/app/components/AddTask.tsx [hunk 12]
```diff
@@ -1,48 +1,40 @@
...
-        setSaveModal(false);
+        setmodalOpened(false);
     };
     const inputChangeConfirmed = () => {
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:34]: save or confirm ?

File: src/app/components/AddTask.tsx [hunk 13]
```diff
@@ -62,7 +54,7 @@ const AddTask = () => {
                         defaultValue={task.name}
                         onChange={handleInputChange}
                     />
-                    <button onClick={openSaveModal} className="p-2 bg-[#2E7D32] text-white rounded">
+                    <button onClick={openmodalOpened} className="p-2 bg-[#2E7D32] text-white rounded">
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:57]: research camel case naming convention

File: src/app/components/Redux/tasks/slice.tsx [hunk 14]
```diff
@@ -1,16 +1,11 @@
 import { PayloadAction, createSlice } from '@reduxjs/toolkit';
-
-type Task = {
-    name: string;
-    id: number;
-    isEditing?: string | null;
-};
+import { TaskType } from '../types';
```
Comment (asapMaki) [inline src/app/components/Redux/tasks/slice.tsx:2]: Task

File: src/app/components/Redux/tasks/slice.tsx [hunk 15]
```diff
@@ -22,16 +17,12 @@ const taskSlice = createSlice({
...
-            newState[editingTaskIndex] = updatedTask;
-            state = newState;
-            return state;
+            state[editingTaskIndex].name = action.payload;
         },
     },
 });
 
-export const { addTask, removeTask, editTask, saveInputChange } = taskSlice.actions;
+export const { addTask, removeTask, editTask, saveTask } = taskSlice.actions;
```
Comment (asapMaki) [inline src/app/components/Redux/tasks/slice.tsx:32]: nice

File: src/app/components/Redux/types.tsx [hunk 16]
```diff
@@ -0,0 +1,9 @@
+export type TaskType = {
+    id: number;
+    name: string;
+    isEditing?: string;
+};
+
+export interface RootState {
```
Comment (asapMaki) [inline src/app/components/Redux/types.tsx:7]: checkout documentation for TS

File: src/app/components/Task.tsx [hunk 17]
```diff
@@ -1,30 +1,26 @@
...
 
     const notConfirmed = () => {
-        setRemoveModal(false);
+        setmodalOpened(false);
     };
     const toggleCheckbox = () => {
         setIsChecked(!isChecked);
     };
-    const handleRemoveTask = (e: Task) => {
+    const handleRemoveTask = (e: TaskType) => {
```
Comment (asapMaki) [inline src/app/components/Task.tsx:23]: handleRemoveTask no need to pass any parameters when you already have task available in scope.
Learn arrow functions

File: src/app/components/Task.tsx [hunk 18]
```diff
@@ -42,9 +38,9 @@ export default function Task({ task }: { task: Task }) {
             <Button className="mr-5 h-[40px]" variant="contained" onClick={() => handleInputChange(task.id)}>
                 Edit
             </Button>
-            {removeModal && (
+            {modalOpened && (
                 <ConfirmModal
-                    openModal={removeModal}
+                    openModal={modalOpened}
                     onConfirm={() => handleRemoveTask(task)}
```
Comment (asapMaki) [inline src/app/components/Task.tsx:44]: onConfirm={handleRemoveTask}

File: src/app/components/TaskList.tsx [hunk 19]
```diff
@@ -1,21 +1,14 @@
...
-interface RootState {
-    tasks: Task[];
-}
 export default function TaskList() {
     const tasks = useSelector((state: RootState) => state.tasks);
     return (
         <>
             <ul>
-                {tasks?.map((task: Task) => (
+                {tasks?.map((task: TaskType) => (
```
Comment (asapMaki) [inline src/app/components/TaskList.tsx:11]: Do you need type when store is already typescript compliant

**Review by asapMaki — CHANGES_REQUESTED** [review]

File: src/app/components/AddTask.tsx [hunk 20]
```diff
@@ -3,15 +3,16 @@ import React, { useEffect, useRef, useState } from 'react';
...
-import { TaskType, RootState } from './Redux/types';
+import { RootState } from './Redux/types';
+import { Task as TaskType } from './Redux/types';
 
 const AddTask = () => {
     const dispatch = useDispatch();
 
     const inputRef = useRef<HTMLInputElement>(null);
-    const [taskName, setTaskName] = useState({ name: '', id: 0 });
+    const [taskName, setTaskName] = useState<TaskType>();
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:13]: is it name (string) or whole Task

File: src/app/components/AddTask.tsx [hunk 21]
```diff
@@ -3,15 +3,16 @@ import React, { useEffect, useRef, useState } from 'react';
...
     const [changedName, setChangedName] = useState<string>('');
-    const [modalOpened, setmodalOpened] = useState<boolean>(false);
+    const [modalOpened, setModalOpened] = useState<boolean>(false);
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:15]: if you don't put false, what value will it be ?

File: src/app/components/AddTask.tsx (hunk 11 above)
Comment (asapMaki) [reply src/app/components/AddTask.tsx:32]: +1

File: src/app/components/Task.tsx [hunk 22]
```diff
@@ -39,11 +39,7 @@ export default function Task({ task }: { task: TaskType }) {
                 Edit
             </Button>
             {modalOpened && (
-                <ConfirmModal
-                    openModal={modalOpened}
-                    onConfirm={() => handleRemoveTask(task)}
-                    onCancel={notConfirmed}
-                />
+                <ConfirmModal openModal={modalOpened} onConfirm={handleRemoveTask} onCancel={notConfirmed} />
```
Comment (asapMaki) [inline src/app/components/Task.tsx:42]: is openModal function or boolean (modalOpened)

**Review by asapMaki — CHANGES_REQUESTED** [review]

File: src/app/components/AddTask.tsx (hunk 20 above)
Comment (limanovic) [reply src/app/components/AddTask.tsx:13]: whole task

File: src/app/components/AddTask.tsx [hunk 23]
```diff
@@ -5,22 +5,24 @@ import { useDispatch, useSelector } from 'react-redux';
...
     const task = tasks.find((task: TaskType) => task.name === task.isEditing);
 
     const handleAddTask = (e: React.FormEvent<HTMLFormElement>) => {
         e.preventDefault();
-        if (!taskName || !taskName.name.trim()) alert('Task cannot be empty');
-        else dispatch(addTask(taskName));
-        setTaskName({ name: '', id: 0 });
+        if (!newTask || !newTask.name.trim()) alert('Task cannot be empty');
+        else dispatch(addTask(newTask));
+        setNewTask({ name: '', id: 0 });
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:25]: in constants, or reducer, create variable emptyTask { name: '', id: 0 }

File: src/app/components/Task.tsx [hunk 24]
```diff
@@ -27,20 +29,31 @@ export default function Task({ task }: { task: TaskType }) {
...
+                    <FontAwesomeIcon icon={faCheck} />
+                </button>
+            ) : (
+                <button
+                    onClick={toggleCheckbox}
+                    className="text-[#1976D2] border border-[#1976D2] p-2 w-[60px] rounded">
+                    Done
+                </button>
             )}
+            {modalOpened && <ConfirmModal openModal={true} onConfirm={handleRemoveTask} onCancel={notConfirmed} />}
```
Comment (asapMaki) [inline src/app/components/Task.tsx:56]: openModal={true}  if it is always true, do you need it ?

**Review by asapMaki — CHANGES_REQUESTED** [review]

**Comment by asapMaki** [comment]
I love the test section, why not make it as a list or something more readable
image

**Comment by asapMaki** [comment]
https://to-do-list-q16cx2q1x-limanovic.vercel.app/

File: src/app/components/TaskList.tsx [hunk 25]
```diff
@@ -1,13 +1,29 @@
...
+    const [modalOpened, setModalOpened] = useState<boolean>();
+    const openModal = () => {
+        setModalOpened(true);
+    };
+    const notConfirmed = () => {
+        setModalOpened(false);
+    };
+    const deleteAllTasks = () => {
+        dispatch(clearTasks());
+        setModalOpened(false);
```
Comment (asapMaki) [inline src/app/components/TaskList.tsx:22]: call closeModal() or as you name it notCOnfiremd

File: src/app/components/TaskList.tsx [hunk 26]
```diff
@@ -16,6 +32,13 @@ export default function TaskList() {
                         <hr />
                     </div>
                 ))}
+                <hr />
+                {tasks.length > 1 && (
```
Comment (asapMaki) [inline src/app/components/TaskList.tsx:36]: tasks.length > 1 - replace with different type of check truthy falsy values

**Review by asapMaki — CHANGES_REQUESTED** [review]

File: src/app/components/Task.tsx [hunk 27]
```diff
@@ -53,7 +53,7 @@ export default function Task({ task }: { task: TaskType }) {
                     Done
                 </button>
             )}
-            {modalOpened && <ConfirmModal openModal onConfirm={handleRemoveTask} onCancel={notConfirmed} />}
+            {openModal && <ConfirmModal onConfirm={handleRemoveTask} onCancel={closeModal} />}
```
Comment (asapMaki) [inline src/app/components/Task.tsx:56]: does this modal have its own animations ? 
With this approach you are blocking those animations

Comment (asapMaki) [reply src/app/components/Task.tsx:56]: Why this instead of passing openModal variable as a prop?

File: src/app/components/TaskList.tsx [hunk 28]
```diff
@@ -1,24 +1,23 @@
...
 import { useState } from 'react';
 import ConfirmModal from './ConfirmModal';
+import { useAppDispatch, useAppSelector } from './Redux/hooks';
 
 export default function TaskList() {
-    const tasks = useSelector((state: RootState) => state.tasks);
-    const dispatch = useDispatch();
+    const tasks = useAppSelector((state) => state.tasks);
+    const dispatch = useAppDispatch();
     const [modalOpened, setModalOpened] = useState<boolean>();
```
Comment (asapMaki) [inline src/app/components/TaskList.tsx:12]: use this var name in component before

**Review by asapMaki — CHANGES_REQUESTED** [review]

File: src/app/components/AddTask.tsx [hunk 29]
```diff
@@ -75,7 +75,11 @@ const AddTask = () => {
                     </button>
                 </form>
             )}
-            {modalOpened && <ConfirmModal onConfirm={inputChangeConfirmed} onCancel={inputChangeNotConfirmed} />}
+            <ConfirmModal
+                modalOpened={modalOpened}
+                onConfirm={inputChangeConfirmed}
```
Comment (asapMaki) [inline src/app/components/AddTask.tsx:80]: still thing inputChangeConfirmed doesn't describe what it actually does

**Review by asapMaki — CHANGES_REQUESTED** [review]

**Review by asapMaki — APPROVED** [review]

**Pull request merged**